app/src/main/java/com/github/featuredetectandroid/MainActivity.kt

```diff
@@ -1,27 +1,153 @@
+                    finish()
                 }
             }
         }
+
+    private fun checkPermissions() {
+        if (allPermissionsGranted()) {
+            startCamera()
+        } else {
+            cameraPermission.launch(Manifest.permission.CAMERA)
+        }
+    }
+    private fun allPermissionsGranted() = REQUIRED_PERMISSIONS.all {
+        ContextCompat.checkSelfPermission(
+            this,
+            it
+        ) == PackageManager.PERMISSION_GRANTED
+    }
+
+    private fun startCamera() {
+        val cameraProviderFuture = ProcessCameraProvider.getInstance(this)
+
+        cameraProviderFuture.addListener({
+            val cameraProvider: ProcessCameraProvider = cameraProviderFuture.get()
+
+            val imageAnalyzer = ImageAnalysis.Builder()
+                .setOutputImageFormat(OUTPUT_IMAGE_FORMAT_YUV_420_888)
+                .build()
+                .also {
+                    it.setAnalyzer(cameraExecutor, PhotoAnalyzer(imageViewModel))
+                }
+
+            val cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA
+
+            try {
+                cameraProvider.unbindAll()
+                cameraProvider.bindToLifecycle(
+                    this,
+                    cameraSelector,
+                    imageAnalyzer
+                )
+            } catch (illegalState: IllegalStateException) {
+                Log.e(TAG, "Use case binding failed: ", illegalState)
+            } catch (illegalArgument: IllegalArgumentException) {
+                Log.e(TAG, "Use case binding failed: ", illegalArgument)
+            }
+        }, ContextCompat.getMainExecutor(this))
+    }
+
+    private class PhotoAnalyzer(val imageViewModel: GrayscaleViewModel) : ImageAnalysis.Analyzer {
+        private fun ByteBuffer.toByteArray(): ByteArray {
+            rewind()
+            val data = ByteArray(remaining())
+            get(data)
+            return data
+        }
+
+        private fun rotateRightOn90Deg(
+            grayscaleByteArray: ByteArray,
+            width: Int,
+            height: Int
+        ) = ByteArray(width * height).apply {
+            for (i in 0 until width) {
+                for (j in 0 until height) {
+                    this[i * height + j] = grayscaleByteArray[i + (height - 1 - j) * width]
+                }
+            }
+        }
+
+        // Here was hardcoded the 90 deg right rotation because of wrong orientation of images got from camera.
+        override fun analyze(image: ImageProxy) {
+            val width = image.width
+            val height = image.height
+            val buffer = image.planes[0].buffer
+            val grayScaleByteArray = buffer.toByteArray()
+            val rotated = rotateRightOn90Deg(grayScaleByteArray, width, height)
+            imageViewModel.setPicture(rotated, height, width)
+            image.close()
+        }
+    }
+
+    companion object {
+        private const val TAG = "FeatureDetectApp"
```

Comment:
`TAG` is ussually the name of the class. It's just a convention but I think debugging is much easier this way -- you immediately see where to look from the logs.

Reply:
fixed